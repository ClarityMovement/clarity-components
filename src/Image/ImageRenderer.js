import React from 'react';
import T from 'prop-types';

import Card from 'material-ui/Card/Card';
import CardActions from 'material-ui/Card/CardActions';
import CardText from 'material-ui/Card/CardText';
import RaisedButton from 'material-ui/RaisedButton';
import FlatButton from 'material-ui/FlatButton';
import LinearProgress from 'material-ui/LinearProgress';

import Img from 'react-image';

import ImageCropper from './ImageCropper';
import ImageSelector from './ImageSelector';

const styles = {
  root: { maxWidth: 300 },
  content: { maxHeight: 300, paddingBottom: 0 },
  cancelButton: { marginRight: 0, minWidth: 85 },
  loader: { marginBottom: 5 },
};

const ImageRenderer = ({
  image,
  scale,
  url,
  cropping,
  failed,
  uploaded,
  uploading,
  onUpload,
  openEditor,
  reset,
  setEditor,
  setImage,
  setScale,
}) => (
  <Card style={styles.root}>
    <CardText style={styles.content}>
      {uploading ? <LinearProgress style={styles.loader} /> : null}
      {cropping ? (
        <ImageCropper image={image} setScale={setScale} setEditor={setEditor} scale={scale} />
      ) : null}
      {uploaded ? <Img width="100%" height="100%" src={url} /> : null}
    </CardText>
    <CardActions>
      {image && !uploaded ? (
        <RaisedButton
          name="Upload"
          secondary
          disabled={uploading}
          label={failed ? 'Retry' : 'Upload'}
          onTouchTap={onUpload}
        />
      ) : null}
      {uploaded ? (
        <RaisedButton name="Crop" primary label="Crop" onTouchTap={openEditor} />
      ) : (
        <RaisedButton
          name="Select"
          primary
          label={image ? 'Change' : 'Select Image'}
          labelPosition="before"
        >
          <ImageSelector setImage={setImage} />
        </RaisedButton>
      )}
      <FlatButton style={styles.cancelButton} disabled={!image} label="Reset" onTouchTap={reset} />
    </CardActions>
  </Card>
);

ImageRenderer.propTypes = {
  image: T.shape({
    preview: T.string.isRequired,
  }),
  scale: T.number,
  url: T.string,
  cropping: T.bool,
  failed: T.bool,
  uploaded: T.bool,
  uploading: T.bool,
  onUpload: T.func,
  openEditor: T.func,
  reset: T.func,
  setEditor: T.func,
  setImage: T.func,
  setScale: T.func,
};

ImageRenderer.defaultProps = {
  image: undefined,
  scale: 1,
  url: undefined,
  cropping: undefined,
  failed: undefined,
  uploaded: undefined,
  uploading: undefined,
  onUpload: undefined,
  openEditor: undefined,
  reset: undefined,
  setEditor: undefined,
  setImage: undefined,
  setScale: undefined,
};

export default ImageRenderer;
