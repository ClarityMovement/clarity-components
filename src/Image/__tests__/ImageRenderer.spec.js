/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';

import ImageRenderer from '../ImageRenderer';

const url = 'https://pics.example.com/200/100';
const image = { preview: url };
const onUpload = jest.fn();
const openEditor = jest.fn();
const reset = jest.fn();
const setImage = jest.fn();
const setScale = jest.fn();
const setEditor = jest.fn();

const defaultProps = {
  onUpload,
  openEditor,
  reset,
  setImage,
  setScale,
  setEditor,
};

describe('<ImageRenderer {...props} />: elements tree', () => {
  test('with default props', () => {
    const wrapper = shallow(<ImageRenderer {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('with cropping props after selected an image', () => {
    const props = {
      image,
      cropping: true,
      scale: 2,
    };
    const wrapper = shallow(<ImageRenderer {...defaultProps} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('with uploading props', () => {
    const props = {
      image,
      cropping: true,
      scale: 2,
      uploading: true,
    };
    const wrapper = shallow(<ImageRenderer {...defaultProps} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('with uploaded props', () => {
    const props = {
      image,
      url,
      uploaded: true,
    };
    const wrapper = shallow(<ImageRenderer {...defaultProps} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('with failed (to upload) props', () => {
    const props = {
      image,
      cropping: true,
      scale: 2,
      failed: true,
    };
    const wrapper = shallow(<ImageRenderer {...defaultProps} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('crop after upload succeed', () => {
    const props = {
      image,
      url,
      cropping: true,
    };
    const wrapper = shallow(<ImageRenderer {...defaultProps} {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
