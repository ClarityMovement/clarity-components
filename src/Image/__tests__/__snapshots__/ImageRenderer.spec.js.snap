// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageRenderer {...props} />: elements tree crop after upload succeed 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  >
    <withHandlers(ImageCropper)
      image={
        Object {
          "preview": "https://pics.example.com/200/100",
        }
      }
      scale={1}
      setEditor={[Function]}
      setScale={[Function]}
    />
  </CardText>
  <CardActions>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Upload"
      labelPosition="after"
      name="Upload"
      onTouchTap={[Function]}
      primary={false}
      secondary={true}
    />
    <RaisedButton
      containerElement="label"
      disabled={false}
      fullWidth={false}
      label="Change"
      labelPosition="before"
      name="Select"
      primary={true}
      secondary={false}
    >
      <withHandlers(ImageSelector)
        setImage={[Function]}
      />
    </RaisedButton>
    <FlatButton
      disabled={false}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;

exports[`<ImageRenderer {...props} />: elements tree with cropping props after selected an image 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  >
    <withHandlers(ImageCropper)
      image={
        Object {
          "preview": "https://pics.example.com/200/100",
        }
      }
      scale={2}
      setEditor={[Function]}
      setScale={[Function]}
    />
  </CardText>
  <CardActions>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Upload"
      labelPosition="after"
      name="Upload"
      onTouchTap={[Function]}
      primary={false}
      secondary={true}
    />
    <RaisedButton
      containerElement="label"
      disabled={false}
      fullWidth={false}
      label="Change"
      labelPosition="before"
      name="Select"
      primary={true}
      secondary={false}
    >
      <withHandlers(ImageSelector)
        setImage={[Function]}
      />
    </RaisedButton>
    <FlatButton
      disabled={false}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;

exports[`<ImageRenderer {...props} />: elements tree with default props 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  />
  <CardActions>
    <RaisedButton
      containerElement="label"
      disabled={false}
      fullWidth={false}
      label="Select Image"
      labelPosition="before"
      name="Select"
      primary={true}
      secondary={false}
    >
      <withHandlers(ImageSelector)
        setImage={[Function]}
      />
    </RaisedButton>
    <FlatButton
      disabled={true}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;

exports[`<ImageRenderer {...props} />: elements tree with failed (to upload) props 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  >
    <withHandlers(ImageCropper)
      image={
        Object {
          "preview": "https://pics.example.com/200/100",
        }
      }
      scale={2}
      setEditor={[Function]}
      setScale={[Function]}
    />
  </CardText>
  <CardActions>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Retry"
      labelPosition="after"
      name="Upload"
      onTouchTap={[Function]}
      primary={false}
      secondary={true}
    />
    <RaisedButton
      containerElement="label"
      disabled={false}
      fullWidth={false}
      label="Change"
      labelPosition="before"
      name="Select"
      primary={true}
      secondary={false}
    >
      <withHandlers(ImageSelector)
        setImage={[Function]}
      />
    </RaisedButton>
    <FlatButton
      disabled={false}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;

exports[`<ImageRenderer {...props} />: elements tree with uploaded props 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  >
    <Img
      height="100%"
      loader={false}
      src="https://pics.example.com/200/100"
      unloader={false}
      width="100%"
    />
  </CardText>
  <CardActions>
    <RaisedButton
      disabled={false}
      fullWidth={false}
      label="Crop"
      labelPosition="after"
      name="Crop"
      onTouchTap={[Function]}
      primary={true}
      secondary={false}
    />
    <FlatButton
      disabled={false}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;

exports[`<ImageRenderer {...props} />: elements tree with uploading props 1`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <CardText
    style={
      Object {
        "maxHeight": 300,
        "paddingBottom": 0,
      }
    }
  >
    <LinearProgress
      max={100}
      min={0}
      mode="indeterminate"
      style={
        Object {
          "marginBottom": 5,
        }
      }
      value={0}
    />
    <withHandlers(ImageCropper)
      image={
        Object {
          "preview": "https://pics.example.com/200/100",
        }
      }
      scale={2}
      setEditor={[Function]}
      setScale={[Function]}
    />
  </CardText>
  <CardActions>
    <RaisedButton
      disabled={true}
      fullWidth={false}
      label="Upload"
      labelPosition="after"
      name="Upload"
      onTouchTap={[Function]}
      primary={false}
      secondary={true}
    />
    <RaisedButton
      containerElement="label"
      disabled={false}
      fullWidth={false}
      label="Change"
      labelPosition="before"
      name="Select"
      primary={true}
      secondary={false}
    >
      <withHandlers(ImageSelector)
        setImage={[Function]}
      />
    </RaisedButton>
    <FlatButton
      disabled={false}
      fullWidth={false}
      label="Reset"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={false}
      secondary={false}
      style={
        Object {
          "marginRight": 0,
          "minWidth": 85,
        }
      }
    />
  </CardActions>
</Card>
`;
